File: c++/332. Reconstruct Itinerary.cpp
class Solution {
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {
        int ne = tickets.size();
        for (auto &t : tickets) {
            graph[t.first][t.second]++;
        }
        
        vector<string> res;
        string s = "JFK";
        path.push_back(s);
        bool found = dfs(s, ne, res);
        graph.clear();
        path.clear();
        return res;
    }
    
private:
    map<string, map<string, int>> graph;
    vector<string> path;
    
    bool dfs(string s, int ec, vector<string> &res) {
        if (ec == 0) {
            res = path;
            return true;
        }
        
        string s1;
        for (auto it = graph[s].begin(); it != graph[s].end(); it++) {
            if (it->second == 0)
                continue;
            s1 = it->first;
            --it->second;
            path.push_back(s1);
            if (dfs(s1, ec-1, res))
                return true;
            ++it->second;
            path.pop_back();
        }
        
        return false;
    }
};

File: python/332. Reconstruct Itinerary.py
#!usr/bin/env python  
#-*- coding:utf-8 _*-  

import collections

class Solution(object):
    def findItinerary(self, tickets):
        """
        # 思路：dfs，遍历的时候要删除之前走过的边
        # time：O(N!), space: O(1)     
        """
        edges = collections.defaultdict(list)
        for start, end in tickets:
            edges[start].append(end)
        
        for start in edges.iterkeys():
            edges[start].sort(reverse=True)
        
        path = []
        def dfs(start):
            while edges[start]:
                dfs(edges[start].pop())
            path.append(start)
        
        dfs('JFK')
        return path[::-1]



if __name__ == '__main__':
    # tickets = [["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]]
    tickets = [["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]]
    su = Solution()
    su.findItinerary2(tickets)


      


            
